powerauth-restful-security-spring/src/main/java/io/getlime/security/powerauth/rest/api/spring/filter/EncryptionResponseBodyAdvice.java

```diff
@@ -86,20 +87,67 @@ public EciesEncryptedResponse beforeBodyWrite(Object response, MethodParameter m
         }
 
         // Convert response to JSON
-        byte[] responseBytes;
         try {
-            final ByteArrayOutputStream baos = new ByteArrayOutputStream();
-            objectMapper.writeValue(baos, response);
-            responseBytes = baos.toByteArray();
+            byte[] responseBytes = serializeResponseObject(response);
 
             // Encrypt response using decryptor and return ECIES cryptogram
             final EciesDecryptor eciesDecryptor = eciesEncryption.getEciesDecryptor();
             EciesCryptogram cryptogram = eciesDecryptor.encryptResponse(responseBytes);
             String encryptedDataBase64 = BaseEncoding.base64().encode(cryptogram.getEncryptedData());
             String macBase64 = BaseEncoding.base64().encode(cryptogram.getMac());
-            return new EciesEncryptedResponse(encryptedDataBase64, macBase64);
+
+            // Convert encrypted response based on response type and return it
+            final EciesEncryptedResponse encryptedResponse = new EciesEncryptedResponse(encryptedDataBase64, macBase64);
+            return convertEncryptedResponse(response.getClass(), encryptedResponse);
         } catch (Exception ex) {
             return null;
         }
     }
+
+    /**
+     * Serialize response object to byte[].
+     *
+     * @param response Response object.
+     * @return Response data as byte[].
+     * @throws IOException In case JSON serialization fails.
+     */
+    private byte[] serializeResponseObject(Object response) throws IOException {
+        if (response.getClass() == byte[].class) {
+            // Response data is raw byte[], data conversion is not required
+            return (byte[]) response;
+        } else {
+            // Convert response object to byte[] using ObjectMapper
+            final ByteArrayOutputStream baos = new ByteArrayOutputStream();
+            objectMapper.writeValue(baos, response);
+            return baos.toByteArray();
+        }
+    }
+
+    /**
+     * Convert encrypted response to response object serializable by Spring based on response type.
+     *
+     * @param responseType Response object type.
+     * @param encryptedResponse Encrypted response to convert.
+     * @return Converted encrypted response.
+     * @throws IOException In case JSON serialization fails.
+     */
+    private Object convertEncryptedResponse(Class<?> responseType, EciesEncryptedResponse encryptedResponse) throws IOException {
+        if (responseType == byte[].class) {
```

Comment:
Good point, I will investigate it.

Reply:
The method is no longer necessary. The problem was that the `ResponseBodyAdvice` was modifying behaviour of all HTTP message converters (including `ByteArrayHttpMessageConverter` and `StringHttpMessageConverter` which were problematic). Once I specified that the advice should be applied only in case when conversion to JSON is performed, the problem went away.

The code change fixing the problem:
```
    public boolean supports(@NonNull MethodParameter methodParameter, @NonNull Class<? extends HttpMessageConverter<?>> converterClass) {
        return methodParameter.hasMethodAnnotation(PowerAuthEncryption.class) && converterClass.isAssignableFrom(MappingJackson2HttpMessageConverter.class);
    }
```

It is also possible to use transparent encryption when using other HTTP message converters, however I suggest we don't implement it, because we don't know the use cases yet (e.g. XML, Gson, ...). Whoever will want to get other format than JSON can autowire their own `ResponseBodyAdvice`.